File: day5/first/main.py
from collections import namedtuple

Sdr = namedtuple('SDR', ['src', 'dst', 'rng'])

seeds = []
maps = [[] for _ in range(7)]

with open("../input.txt") as file:
    seeds = map(int, file.readline().lstrip("seeds:").strip().split())

    idx = -1
    while (line := file.readline()):
        line = line.strip()
        if line == "":
            continue
        if not line[0].isnumeric():
            idx += 1
            continue 
        
        dsr = list(map(int, line.split()))
        maps[idx].append(Sdr(dsr[1], dsr[0], dsr[2]))

closest = float('inf')

for seed in seeds:
    for map in maps:
        for sdr in map:
            if sdr.src <= seed <= sdr.src+sdr.rng:
                seed = sdr.dst + (seed - sdr.src)
                break
    closest = min(closest, seed)

print(closest)

File: day5/first/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
	"unicode"
)

type SDR struct {
	src, dst, rng int
}

func main() {
	input, err := os.Open("../input.txt")
	if err != nil {
		os.Exit(1)
	}
	defer input.Close()
	scanner := bufio.NewScanner(input)

	maps := make([][]SDR, 7)

	scanner.Scan()
	seeds := atoi(strings.TrimPrefix(scanner.Text(), "seeds:"))

	mapIdx := -1
	for scanner.Scan() {
		text := scanner.Text()
		if text == "" {
			continue
		}
		if !unicode.IsDigit(rune(text[0])) {
			mapIdx++
			continue
		}
		dsr := atoi(text)
		maps[mapIdx] = append(maps[mapIdx], SDR{dsr[1], dsr[0], dsr[2]})
	}

	closest := math.MaxInt

	for _, seed := range seeds {
		for _, m := range maps {
			for _, sdr := range m {
				if seed >= sdr.src && seed <= sdr.src+sdr.rng {
					seed = sdr.dst + (seed - sdr.src)
					break
				}
			}
		}
		closest = min(closest, seed)
	}

	fmt.Println(closest)
}

func atoi(str string) []int {
	strSlice := strings.Split(strings.TrimSpace(str), " ")
	intSlice := []int{}
	for _, e := range strSlice {
		num, _ := strconv.Atoi(e)
		intSlice = append(intSlice, num)
	}
	return intSlice
}
